lib/storage/src/content_manager/consensus_manager.rs

```diff
@@ -194,14 +190,90 @@ impl<C: CollectionContainer> ConsensusManager<C> {
     }
 
     pub fn first_voter(&self) -> PeerId {
-        match self.first_voter.read().as_ref() {
-            Some(id) => *id,
-            None => self.this_peer_id(),
+        let state = self.persistent.read();
+
+        match state.first_voter() {
+            Some(peer_id) if peer_id != PeerId::MAX => peer_id,
+            _ => state.this_peer_id(),
         }
     }
 
-    pub fn set_first_voter(&self, id: PeerId) {
-        *self.first_voter.write() = Some(id);
+    pub fn set_first_voter(&self, id: PeerId) -> Result<(), StorageError> {
+        self.persistent.write().set_first_voter(id)
+    }
+
+    pub fn recover_first_voter(&self) -> Result<(), StorageError> {
+        if self.persistent.read().first_voter().is_none() {
+            if let Some(peer_id) = self.recover_first_voter_impl()? {
+                self.set_first_voter(peer_id)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn recover_first_voter_impl(&self) -> Result<Option<PeerId>, StorageError> {
+        let wal = self.wal.lock();
+
+        let Some(first_entry) = wal.first_entry()? else {
+            log::warn!("TODO: empty WAL");
+            return Ok(None);
+        };
+
+        let Some(last_entry) = wal.last_entry()? else {
+            log::warn!("TODO: invalid WAL (no last entry)");
+            return Ok(None);
+        };
+
+        if first_entry.index != 1 {
+            log::warn!("TODO: compacted WAL");
+            return Ok(Some(PeerId::MAX));
+        }
+
+        let mut peers: HashSet<_> = self.peers().into_iter().collect();
+
+        for index in first_entry.index..last_entry.index + 1 {
```

Comment:
I was a bit confused by this at first. Maybe we can add some comment to describe it.

```suggestion
        // Drain added peers from list, if we have one left that'd be our origin peer
        let mut peers: HashSet<_> = self.peers().into_iter().collect();
        for index in first_entry.index..last_entry.index + 1 {
```

I figured, maybe we can do it the other way around. To find a remove for which we don't have a add entry. But I assume you're not doing that because we also store garbage removes in the WAL for peers that never existed.

Reply:
Added a comment.